File: src/tools.js
/**
 * Tools available to the agent.
 * @module tools
 */

import { exec } from "child_process";
import { promisify } from "util";
import fs from "fs/promises";
import path from "path";
import { limitOutput } from "./utils.js";
import { callLLM } from "./llm.js";
import { config } from "./config.js";

const execAsync = promisify(exec);

/**
 * Executes a bash command.
 */
async function bash({ command }) {
  try {
    // Added timeout of 30 seconds
    const { stdout, stderr } = await execAsync(command, {
      maxBuffer: 10 * 1024 * 1024,
      timeout: 30000,
    });
    let output = stdout + (stderr ? `\nSTDERR:\n${stderr}` : "");

    if (!output.trim()) {
      return "(Command executed successfully with no output)";
    }
    return limitOutput(output);
  } catch (error) {
    return limitOutput(
      `Error executing command:\n${error.message}\nSTDERR:\n${
        error.stderr || ""
      }`,
      true,
    );
  }
}

/**
 * Search for files by name pattern, respecting .gitignore.
 */
async function search_files({ pattern }) {
  if (!pattern) return "Error: 'pattern' is required.";
  return await bash({ command: `fd "${pattern}"` });
}

/**
 * Search for a string in files, respecting .gitignore.
 */
async function search_string({ query }) {
  if (!query) return "Error: 'query' is required.";
  // Explicitly search the current directory to ensure it doesn't wait for stdin
  // -- is used to stop argument parsing, ensuring 'query' isn't interpreted as a flag
  return await bash({ command: `rg -n -C 5 -- "${query}" .` });
}

/**
 * Reads a file or a specific line range.
 * Enforces a maximum of 500 lines per read.
 */
async function read_file({ path: filePath, start_line, end_line }) {
  if (!filePath) return "Error: 'path' is required.";
  try {
    const content = await fs.readFile(filePath, "utf8");
    const lines = content.split("\n");

    let start = 0;
    let end = lines.length;

    if (start_line !== undefined) {
      start = parseInt(start_line) - 1; // 1-based to 0-based
      if (isNaN(start) || start < 0) start = 0;
    }

    if (end_line !== undefined) {
      end = parseInt(end_line);
      if (isNaN(end)) end = lines.length;
    }

    // Enforce 500 line limit
    const MAX_LINES = 500;

    if (end - start > MAX_LINES) {
      end = start + MAX_LINES;
    }

    if (start >= lines.length) return "";
    if (end > lines.length) end = lines.length;
    if (start < 0) start = 0;

    const selectedLines = lines.slice(start, end);

    return limitOutput(
      `(Total lines: ${lines.length})\n` + selectedLines.join("\n"),
    );
  } catch (error) {
    return `Error reading file: ${error.message}`;
  }
}

/**
 * Updates (or creates) a file.
 * Supports full overwrite or partial search-and-replace.
 */
async function update_file({ path: filePath, content, old_content }) {
  if (!filePath || content === undefined)
    return "Error: 'path' and 'content' are required.";

  try {
    if (old_content) {
      // Partial update mode
      let currentFileContent;
      try {
        currentFileContent = await fs.readFile(filePath, "utf8");
      } catch (err) {
        return `Error reading file for partial update: ${err.message}. (File must exist for partial updates)`;
      }

      if (!currentFileContent.includes(old_content)) {
        return "Error: 'old_content' text block not found in file. Please ensure exact match (including whitespace/indentation).";
      }

      const newFileContent = currentFileContent.replace(old_content, content);

      await fs.writeFile(filePath, newFileContent, "utf8");
      return `Successfully updated ${filePath} (partial replace).`;
    } else {
      // Full overwrite mode
      await fs.mkdir(path.dirname(filePath), { recursive: true });
      await fs.writeFile(filePath, content, "utf8");
      return `Successfully updated ${filePath} (full overwrite).`;
    }
  } catch (error) {
    return `Error updating file: ${error.message}`;
  }
}

/**
 * Search the web using Google Search Grounding.
 * Use this to get up-to-date information from the internet.
 */
async function google_search({ query }) {
  if (!query) return "Error: 'query' is required.";

  try {
    const searchMessages = [
      { role: "user", content: `Perform a google search for: ${query}` },
    ];
    let tools = [];

    tools = [
      {
        type: "function",
        function: {
          name: "__google_search_trigger__",
          description: "Trigger Google Search Grounding",
          parameters: { type: "object", properties: {} },
        },
      },
    ];

    const { message } = await callLLM(
      searchMessages,
      tools,
      "google/gemini-3-pro-preview",
    );
    return message.content || "No results found.";
  } catch (error) {
    return `Error performing Google Search: ${error.message}`;
  }
}

/**
 * Saves an image from the clipboard to a temporary file.
 * Returns the path to the temporary file, or null if failed.
 */
export async function saveClipboardImage() {
  const tempFilePath = path.join(
    "/tmp",
    `clipboard_${Date.now()}_${Math.random().toString(36).slice(2)}.png`,
  );
  try {
    // Try wl-paste first (Wayland)
    await execAsync(`wl-paste -t image/png > "${tempFilePath}"`);
    return tempFilePath;
  } catch (err) {
    try {
      // Try xclip (X11)
      await execAsync(
        `xclip -selection clipboard -t image/png -o > "${tempFilePath}"`,
      );
      return tempFilePath;
    } catch (err2) {
      // Try pngpaste (MacOS)
      try {
        await execAsync(`pngpaste "${tempFilePath}"`);
        return tempFilePath;
      } catch (err3) {
        return null;
      }
    }
  }
}

/**
 * Describes an image using Gemini 3 Pro Preview.
 */
async function describe_image({ path: singlePath, paths }) {
  if (!singlePath && (!paths || paths.length === 0)) {
    return "Error: 'path' or 'paths' is required.";
  }

  const imagePaths = paths || [singlePath];
  const tempFiles = [];
  try {
    const content = [
      {
        type: "text",
        text: "Describe these images in detail. Identify objects, colors, text, and the overall scene.",
      },
    ];

    for (let i = 0; i < imagePaths.length; i++) {
      let imagePath = imagePaths[i];

      if (imagePath === "clipboard") {
        const tempFilePath = await saveClipboardImage();
        if (!tempFilePath) {
          return `Error reading from clipboard: Could not find wl-paste, xclip, or pngpaste, or failed to get image.`;
        }
        tempFiles.push(tempFilePath);
        imagePath = tempFilePath;
      }

      const imageBuffer = await fs.readFile(imagePath);
      const base64Image = imageBuffer.toString("base64");
      const ext = path.extname(imagePath).toLowerCase();

      let mimeType = "image/jpeg";
      if (ext === ".png") mimeType = "image/png";
      if (ext === ".webp") mimeType = "image/webp";
      if (ext === ".heic") mimeType = "image/heic";
      if (ext === ".heif") mimeType = "image/heif";

      content.push({
        type: "image_url",
        image_url: {
          url: `data:${mimeType};base64,${base64Image}`,
        },
      });
    }

    const messages = [{ role: "user", content: content }];
    const { message } = await callLLM(
      messages,
      [],
      "google/gemini-3-pro-preview",
    );

    return message.content || "Error: No description returned.";
  } catch (error) {
    return `Error describing image: ${error.message}`;
  } finally {
    for (const tempFile of tempFiles) {
      await fs.unlink(tempFile).catch(() => {});
    }
  }
}

export const tools = [
  {
    type: "function",
    function: {
      name: "bash",
      description:
        "Execute a bash command. Use this for all system operations, file manipulation, and information retrieval.",
      parameters: {
        type: "object",
        properties: {
          command: {
            type: "string",
            description: "The bash command to execute.",
          },
        },
        required: ["command"],
      },
    },
  },
  {
    type: "function",
    function: {
      name: "search_files",
      description: "Search for files by name pattern, respecting .gitignore.",
      parameters: {
        type: "object",
        properties: {
          pattern: {
            type: "string",
            description: "The filename pattern to search for.",
          },
        },
        required: ["pattern"],
      },
    },
  },
  {
    type: "function",
    function: {
      name: "search_string",
      description:
        "Search for a string in files, respecting .gitignore. Returns 5 lines of context around matches.",
      parameters: {
        type: "object",
        properties: {
          query: {
            type: "string",
            description: "The string to search for.",
          },
        },
        required: ["query"],
      },
    },
  },
  {
    type: "function",
    function: {
      name: "read_file",
      description: "Read the content of a file. Returns up to 500 lines.",
      parameters: {
        type: "object",
        properties: {
          path: {
            type: "string",
            description: "The path to the file.",
          },
          start_line: {
            type: "integer",
            description: "Start line number (1-based, inclusive).",
          },
          end_line: {
            type: "integer",
            description: "End line number (1-based, inclusive).",
          },
        },
        required: ["path"],
      },
    },
  },
  {
    type: "function",
    function: {
      name: "update_file",
      description:
        "Update a file. You can overwrite the entire file or perform a partial find-and-replace.",
      parameters: {
        type: "object",
        properties: {
          path: {
            type: "string",
            description: "The path to the file.",
          },
          content: {
            type: "string",
            description: "The new content to write or to substitute.",
          },
          old_content: {
            type: "string",
            description:
              "Optional. If provided, this specific text block will be replaced by 'content' in the file. If omitted, the entire file is overwritten.",
          },
        },
        required: ["path", "content"],
      },
    },
  },
  {
    type: "function",
    function: {
      name: "google_search",
      description:
        "Perform a web search using Google Search Grounding. Use this to get up-to-date information from the internet.",
      parameters: {
        type: "object",
        properties: {
          query: {
            type: "string",
            description: "The search query.",
          },
        },
        required: ["query"],
      },
    },
  },
  {
    type: "function",
    function: {
      name: "describe_image",
      description:
        "Describe one or more images. Returns a detailed text description. Can read from clipboard by passing 'clipboard' as one of the paths.",
      parameters: {
        type: "object",
        properties: {
          paths: {
            type: "array",
            items: {
              type: "string",
            },
            description:
              "Array of paths to image files, or 'clipboard' to read from system clipboard.",
          },
        },
        required: ["paths"],
      },
    },
  },
];

export const toolImplementations = {
  bash,
  search_files,
  search_string,
  read_file,
  update_file,
  google_search,
  describe_image,
};

File: src/tools.py
"""Tool implementations and schemas for the agent."""

import subprocess
import os
import time
import random
import base64
from .utils import truncate_output

# ---------------------------------------------------------------------------
# Tool Implementations
# ---------------------------------------------------------------------------


async def bash(command: str) -> str:
    """Execute a bash command with timeout."""
    try:
        proc = subprocess.run(command, shell=True, capture_output=True, text=True, timeout=30)
        output = proc.stdout + (f"\nSTDERR:\n{proc.stderr}" if proc.stderr else "")
        return truncate_output(output) if output.strip() else "(Command executed successfully with no output)"
    except subprocess.TimeoutExpired:
        return "Error: Command timed out after 30 seconds."
    except Exception as e:
        return f"Error executing command: {e}"


async def search_files(pattern: str) -> str:
    """Search for files by name pattern using fd."""
    if not pattern:
        return "Error: 'pattern' is required."
    return await bash(f'fd "{pattern}"')


async def search_string(query: str) -> str:
    """Search for a string in files using ripgrep."""
    if not query:
        return "Error: 'query' is required."
    return await bash(f'rg -n -C 5 -- "{query}" .')


async def read_file(path: str, start_line: int = None, end_line: int = None) -> str:
    """Read file content with optional line range."""
    if not path:
        return "Error: 'path' is required."
    try:
        with open(path, "r", encoding="utf-8") as f:
            lines = f.readlines()

        start = max(0, (start_line or 1) - 1)
        end = min(len(lines), end_line or len(lines))

        # Limit to 500 lines max
        if end - start > 500:
            end = start + 500

        return truncate_output(f"(Total lines: {len(lines)})\n" + "".join(lines[start:end]))
    except Exception as e:
        return f"Error reading file: {e}"


async def update_file(path: str, content: str, old_content: str = None) -> str:
    """Update a file (full overwrite or partial replace)."""
    if not path or content is None:
        return "Error: 'path' and 'content' are required."
    try:
        if old_content:
            with open(path, "r", encoding="utf-8") as f:
                current = f.read()
            if old_content not in current:
                return "Error: 'old_content' text block not found in file. Ensure exact match (including whitespace)."
            content = current.replace(old_content, content)

        os.makedirs(os.path.dirname(path) or ".", exist_ok=True)
        with open(path, "w", encoding="utf-8") as f:
            f.write(content)
        return f"Successfully updated {path}."
    except Exception as e:
        return f"Error updating file: {e}"


async def save_clipboard_image() -> str | None:
    """Save clipboard image to temp file. Returns path or None."""
    temp_path = f"/tmp/clipboard_{int(time.time())}_{random.randint(1000, 9999)}.png"
    commands = [
        f'wl-paste -t image/png > "{temp_path}"',
        f'xclip -selection clipboard -t image/png -o > "{temp_path}"',
        f'pngpaste "{temp_path}"',
    ]
    for cmd in commands:
        try:
            subprocess.run(cmd, shell=True, check=True, capture_output=True)
            if os.path.exists(temp_path) and os.path.getsize(temp_path) > 0:
                return temp_path
        except subprocess.CalledProcessError:
            continue
    return None


async def describe_image(paths: list) -> str:
    """Describe images using LLM vision."""
    from .llm import call_llm

    if not paths:
        return "Error: 'paths' is required."

    image_paths = paths if isinstance(paths, list) else [paths]
    content = [{"type": "text", "text": "Describe these images in detail."}]

    for p in image_paths:
        actual_path = await save_clipboard_image() if p == "clipboard" else p
        if not actual_path:
            return "Error reading from clipboard."
        try:
            with open(actual_path, "rb") as f:
                b64 = base64.b64encode(f.read()).decode("utf-8")
            content.append({"type": "image_url", "image_url": {"url": f"data:image/png;base64,{b64}"}})
        except Exception as e:
            return f"Error reading image {p}: {e}"

    response = await call_llm([{"role": "user", "content": content}], [])
    return response["message"]["content"]


async def google_search(query: str) -> str:
    """Perform a web search using Google Search grounding."""
    from .llm import call_llm

    if not query:
        return "Error: 'query' is required."

    messages = [
        {"role": "system", "content": "Search the web and provide a detailed answer."},
        {"role": "user", "content": query},
    ]
    # Special trigger tool to enable grounding in providers
    tools = [
        {
            "type": "function",
            "function": {
                "name": "__google_search_trigger__",
                "description": "Trigger search",
                "parameters": {"type": "object", "properties": {}},
            },
        }
    ]

    try:
        response = await call_llm(messages, tools)
        return response["message"]["content"]
    except Exception as e:
        return f"Error performing google search: {e}"


# ---------------------------------------------------------------------------
# Tool Registry
# ---------------------------------------------------------------------------

TOOLS = {
    "bash": bash,
    "search_files": search_files,
    "search_string": search_string,
    "read_file": read_file,
    "update_file": update_file,
    "google_search": google_search,
    "describe_image": describe_image,
}

# ---------------------------------------------------------------------------
# Tool Schemas (OpenAI-compatible format)
# ---------------------------------------------------------------------------

TOOL_SCHEMAS = [
    {
        "type": "function",
        "function": {
            "name": "bash",
            "description": "Execute a bash command.",
            "parameters": {
                "type": "object",
                "properties": {"command": {"type": "string", "description": "The bash command to execute."}},
                "required": ["command"],
            },
        },
    },
    {
        "type": "function",
        "function": {
            "name": "search_files",
            "description": "Search for files by name pattern.",
            "parameters": {
                "type": "object",
                "properties": {"pattern": {"type": "string", "description": "The filename pattern to search for."}},
                "required": ["pattern"],
            },
        },
    },
    {
        "type": "function",
        "function": {
            "name": "search_string",
            "description": "Search for a string in files.",
            "parameters": {
                "type": "object",
                "properties": {"query": {"type": "string", "description": "The string to search for."}},
                "required": ["query"],
            },
        },
    },
    {
        "type": "function",
        "function": {
            "name": "read_file",
            "description": "Read the content of a file.",
            "parameters": {
                "type": "object",
                "properties": {
                    "path": {"type": "string", "description": "The path to the file."},
                    "start_line": {"type": "integer", "description": "Start line number."},
                    "end_line": {"type": "integer", "description": "End line number."},
                },
                "required": ["path"],
            },
        },
    },
    {
        "type": "function",
        "function": {
            "name": "update_file",
            "description": "Update a file.",
            "parameters": {
                "type": "object",
                "properties": {
                    "path": {"type": "string", "description": "The path to the file."},
                    "content": {"type": "string", "description": "The new content."},
                    "old_content": {"type": "string", "description": "Optional text block to replace."},
                },
                "required": ["path", "content"],
            },
        },
    },
    {
        "type": "function",
        "function": {
            "name": "google_search",
            "description": "Perform a web search using Google Search Grounding.",
            "parameters": {
                "type": "object",
                "properties": {"query": {"type": "string", "description": "The search query."}},
                "required": ["query"],
            },
        },
    },
    {
        "type": "function",
        "function": {
            "name": "describe_image",
            "description": "Describe one or more images.",
            "parameters": {
                "type": "object",
                "properties": {
                    "paths": {
                        "type": "array",
                        "items": {"type": "string"},
                        "description": "Array of paths to image files, or 'clipboard'.",
                    }
                },
                "required": ["paths"],
            },
        },
    },
]
